src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncExceptionHandlerRewriter.cs

```diff
@@ -350,11 +363,32 @@ private BoundStatement UnpendBranches(
                     }
                 }
 
-                var caseStatement = _F.SwitchSection(i, unpendReturn);
-                cases.Add(caseStatement);
+                if (endOfBlockIsReachable)
+                {
+                    var caseStatement = _F.SwitchSection(i, unpendReturn);
+                    cases.Add(caseStatement);
+                }
+                else
+                {
+                    unconditionalReturn = unpendReturn;
+                }
             }
 
-            return _F.Switch(_F.Local(pendingBranchVar), cases.ToImmutableAndFree());
+            if (!endOfBlockIsReachable && unconditionalReturn is null && !_F.CurrentFunction.IsIterator)
```

Comment:
>IsIterator [](http://example.com/codeflow?start=93&length=10)

From offline discussion, I'll need some more time to understand this iterator scenario. Let's sync tomorrow